### DISCUSSION TITLE
Unable to use nhibernate and Azure SQL Database

### DISCUSSION THREAD

**Discussion by delucaezequiel** [nhibernate/nhibernate-core#2533] [General]
We are trying to connect to an Azure SQL Database with a Client ID / SPN and Password for that purpose we need to set on the connection string the following: `Authentication = "Active Directory Password"`  however it is not supported. When we attempt to use it the following exception is throw 

> System.ArgumentException: 'Keyword not supported: 'authentication'.

Based on microsoft feedback we need to set the access token in the ADO.NET connection, however we couldn't find a NHibernate ADO.NET to set Ex: "connection.access_token" we check Table 3.1. NHibernate ADO.NET Properties.
If we try to skip the configuration in order to use an user provided ADO.NET connection we get an exception for each property that is missing

> System.InvalidOperationException: 'Could not find the dialect in the configuration'
> NHibernate.HibernateException: 'Could not find connection string setting (set connection.connection_string or connection.connection_string_name property)'
> 

the exceptions occurs when we try to build the session factory.

```
public class HibernateUtil
    {
        private static ISessionFactory _sessionFactory;
        private static ISessionFactory SessionFactory
        {
            get
            {
                if (_sessionFactory == null)
                {
                    Configuration objConfiguration = new Configuration();
                    /*
                     * Must define dialect or following exception is thrown
                     * System.InvalidOperationException: 'Could not find the dialect in the configuration'
                     */
                    objConfiguration.SetProperty("dialect", "NHibernate.Dialect.MsSqlAzure2008Dialect");
                    /*
                     * Must define connection_string or following exception is thrown
                     * NHibernate.HibernateException: 'Could not find connection string setting (set connection.connection_string or connection.connection_string_name property)'
                     */
                    objConfiguration.SetProperty("connection.connection_string", "connection string");
                    //Add all Entities / Beans / Domains
                    objConfiguration.AddAssembly(typeof(Entity / Bean / Domain).Assembly);
                    _sessionFactory = objConfiguration.BuildSessionFactory();
                }
                return _sessionFactory;
            }
        }
        public static ISession OpenSession()
        {
            //For user defined ADO.NET connection we are trying to replace this
            return SessionFactory.OpenSession();
           //with this
            return SessionFactory.WithOptions().Connection((new SQLConnection()).get()).OpenSession();
        }
    }
```

**Comment by fredericDelaporte** [comment]
The dialect property is mandatory anyway, even with user supplied connection.
The connection string could be considered optional due to the possibility of supplying connection at session opening, but in most cases, not having it is a mistake. Using user supplied connection on every session is not an usual practice. So you have to provide one, even if it will be an invalid one for your setup. It will not prevent you from using user supplied connection.

But for your case it seems to me you should instead implement your own connection provider (likely deriving from [`DriverConnectionProvider`](https://github.com/nhibernate/nhibernate-core/blob/master/src/NHibernate/Connection/DriverConnectionProvider.cs)) and use the [`connection.provider`](https://nhibernate.info/doc/nhibernate-reference/session-configuration.html#configuration-hibernatejdbc) setting to tell NHibernate to use it.
